File: examples/communication-sdk/android/NeuraTalkCommunicationClient.kt
package com.neuratalk.communication

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

class NeuraTalkCommunicationClient(
    private val apiBaseUrl: String,
    private val apiKey: String,
    private val httpClient: OkHttpClient = OkHttpClient(),
) {
    fun createCallSession(payload: JSONObject): JSONObject {
        val request = Request.Builder()
            .url("$apiBaseUrl/api/communication/create-call-session")
            .addHeader("X-API-Key", apiKey)
            .post(payload.toString().toRequestBody("application/json".toMediaType()))
            .build()

        httpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                error("create-call-session failed: ${response.code}")
            }
            return JSONObject(response.body!!.string())
        }
    }

    fun updateStatus(sessionId: String, state: String, participantIdentity: String): JSONObject {
        val payload = JSONObject()
            .put("state", state)
            .put("participantIdentity", participantIdentity)

        val request = Request.Builder()
            .url("$apiBaseUrl/api/communication/call-status/$sessionId")
            .addHeader("X-API-Key", apiKey)
            .post(payload.toString().toRequestBody("application/json".toMediaType()))
            .build()

        httpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("status update failed: ${response.code}")
            return JSONObject(response.body!!.string())
        }
    }

    fun activateFallback(sessionId: String): JSONObject {
        val request = Request.Builder()
            .url("$apiBaseUrl/api/communication/sessions/$sessionId/fallback")
            .addHeader("X-API-Key", apiKey)
            .post("{}".toRequestBody("application/json".toMediaType()))
            .build()

        httpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("fallback failed: ${response.code}")
            return JSONObject(response.body!!.string())
        }
    }
}

File: examples/communication-sdk/ios/NeuraTalkCommunicationClient.swift
import Foundation

final class NeuraTalkCommunicationClient {
    private let apiBaseURL: URL
    private let apiKey: String
    private let session: URLSession

    init(apiBaseURL: URL, apiKey: String, session: URLSession = .shared) {
        self.apiBaseURL = apiBaseURL
        self.apiKey = apiKey
        self.session = session
    }

    func createCallSession(payload: [String: Any]) async throws -> [String: Any] {
        try await postJSON(path: "/api/communication/create-call-session", payload: payload)
    }

    func updateStatus(sessionId: String, state: String, participantIdentity: String) async throws -> [String: Any] {
        try await postJSON(
            path: "/api/communication/call-status/\(sessionId)",
            payload: [
                "state": state,
                "participantIdentity": participantIdentity
            ]
        )
    }

    func activateFallback(sessionId: String) async throws -> [String: Any] {
        try await postJSON(path: "/api/communication/sessions/\(sessionId)/fallback", payload: [:])
    }

    private func postJSON(path: String, payload: [String: Any]) async throws -> [String: Any] {
        var request = URLRequest(url: apiBaseURL.appendingPathComponent(path))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, 200..<300 ~= httpResponse.statusCode else {
            throw NSError(domain: "NeuraTalkCommunication", code: 1, userInfo: [NSLocalizedDescriptionKey: "Request failed"])
        }

        return (try JSONSerialization.jsonObject(with: data) as? [String: Any]) ?? [:]
    }
}
